### DISCUSSION TITLE
resteasy-core 6.2.7.Final reports NoClassDefFoundError: javax/ws/rs/ext/WriterInterceptor

### DISCUSSION THREAD

**Discussion by noiter** [resteasy/resteasy#4370] [Q&A]
Hello folks,

I recently started facing a really weird issue when I use `keycloak-admin-client` library which has `resteasy 6.2.7.Final` dependencies:
```
- org.jboss.resteasy:resteasy-client:6.2.7.Final
- org.jboss.resteasy:resteasy-client-api:6.2.7.Final
- org.jboss.resteasy:resteasy-core:6.2.7.Final
- org.jboss.resteasy:resteasy-core-spi:6.2.7.Final
- org.jboss.resteasy:resteasy-jackson2-provider:6.2.7.Final
- org.jboss.resteasy:resteasy-jaxb-provider:6.2.7.Final
- org.jboss.resteasy:resteasy-multipart-provider:6.2.7.Final
```
. When I start up the project, the Keycloak class initialization failed due to blow error, 
```
Caused by: java.lang.NoClassDefFoundError: javax/ws/rs/ext/WriterInterceptor
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProviderInstance(ResteasyProviderFactoryImpl.java:924)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProviderInstance(ResteasyProviderFactoryImpl.java:909)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.register(ResteasyProviderFactoryImpl.java:1373)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.register(ResteasyProviderFactoryImpl.java:119)
	at org.jboss.resteasy.client.jaxrs.internal.ClientConfiguration.register(ClientConfiguration.java:179)
	at org.jboss.resteasy.client.jaxrs.internal.ClientWebTarget.register(ClientWebTarget.java:418)
	at org.jboss.resteasy.client.jaxrs.internal.ClientWebTarget.register(ClientWebTarget.java:26)
	at org.keycloak.admin.client.token.TokenManager.<init>(TokenManager.java:59)
	at org.keycloak.admin.client.Keycloak.<init>(Keycloak.java:91)
	at org.keycloak.admin.client.KeycloakBuilder.build(KeycloakBuilder.java:152)
``` 

and more errors when loading builtin provider but got `NoClassDefFoundError`: `javax/ws/rs/ext/WriterInterceptor`
```
WARN  main org.jboss.resteasy.resteasy_jaxrs.i18n   : RESTEASY002145: NoClassDefFoundError: Unable to load builtin provider org.jboss.resteasy.plugins.providers.multipart.MultipartWriter from jar:file:/app/libs/resteasy-multipart-provider-6.2.7.Final.jar!/META-INF/services/jakarta.ws.rs.ext.Providersjava.lang.NoClassDefFoundError: javax/ws/rs/ext/WriterInterceptor
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProvider(ResteasyProviderFactoryImpl.java:777)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProvider(ResteasyProviderFactoryImpl.java:763)

WARN  main org.jboss.resteasy.resteasy_jaxrs.i18n   : RESTEASY002145: NoClassDefFoundError: Unable to load builtin provider org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider from jar:file:/app/libs/resteasy-jackson2-provider-6.2.7.Final.jar!/META-INF/services/jakarta.ws.rs.ext.Providersjava.lang.NoClassDefFoundError: javax/ws/rs/ext/WriterInterceptor

WARN  main org.jboss.resteasy.resteasy_jaxrs.i18n   : RESTEASY002145: NoClassDefFoundError: Unable to load builtin provider org.jboss.resteasy.plugins.providers.multipart.MultipartFormAnnotationReader from jar:file:/app/libs/resteasy-multipart-provider-6.2.7.Final.jar!/META-INF/services/jakarta.ws.rs.ext.Providersjava.lang.NoClassDefFoundError: javax/ws/rs/ext/WriterInterceptor
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProvider(ResteasyProviderFactoryImpl.java:777)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.registerProvider(ResteasyProviderFactoryImpl.java:763)

...

```

I think what looks weird to me is it is looking for `javax/ws/rs/ext/WriterInterceptor` which can't be found in my project dependency list instead of `jakarta/ws/rs/ext/WriterInterceptor` which obviously exists in the dependency list.

I searched on Internet but found no any related answers, so I am just posting here in case any of you have any idea why this is happening. Thanks in advance.

**Comment by jamezp** [comment]
Check the `mvn dependency:tree -Dincludes=org.jboss.resteasy*`. I wonder if something is overriding a version of RESTEasy.

**Reply by noiter** [reply]
Thanks for your reply. Here is what I get by running `./mvnw dependency:tree -Dincludes=org.jboss.resteasy -pl :application`, `application` is the maven module has dependency on `keycloak-admin-client`.
```
[INFO] \- org.keycloak:keycloak-admin-client:jar:24.0.5:compile
[INFO]    +- org.jboss.resteasy:resteasy-client:jar:6.2.7.Final:compile
[INFO]    |  +- org.jboss.resteasy:resteasy-client-api:jar:6.2.7.Final:compile
[INFO]    |  +- org.jboss.resteasy:resteasy-core-spi:jar:6.2.7.Final:compile
[INFO]    |  \- org.jboss.resteasy:resteasy-core:jar:6.2.7.Final:compile
[INFO]    +- org.jboss.resteasy:resteasy-multipart-provider:jar:6.2.7.Final:compile
[INFO]    +- org.jboss.resteasy:resteasy-jackson2-provider:jar:6.2.7.Final:compile
[INFO]    \- org.jboss.resteasy:resteasy-jaxb-provider:jar:6.2.7.Final:compile
```

It looks fine to me.

**Reply by jamezp** [reply]
Are you deploying to a container? There's nothing in RESTEasy 6.2.7 bringing in any `javax` based dependencies.

**Reply by noiter** [reply]
Yeah I am deploying the app to a container. Yeah I guess it could be something out of the application but the runtime environment. I will check further that if you also don't think this has anything to do with reseteasy itself.

**Reply by jamezp** [reply]
It's not a Jakarta EE 8 container is it? That's the only thing I could think of.

**Reply by noiter** [reply]
No it's not, it is amazoncorretto:17-alpine-jdk based image. I don't see any Jakarta related setup. Guess I will have to dig deeper into the runtime.

**Reply by jamezp** [reply]
My apologies, I wasn't real clear. What I meant was, are you using a standalone Keycloak or deploying it to a container like WildFly or JBoss EAP.

**Reply by noiter** [reply]
I see, all good. It is a Spring boot 3 application having `keycloak-admin-client` dependency. Shouldn't have anything to do with the quarkus-based Keycloak server (v24) which is separately deployed.

**Reply by jamezp** [reply]
Hmm... ...Spring Boot 3 should be supporting the jakarta namespace. It really seems like something is bringing in an older version of RESTEasy. I don't know much about Spring Boot, but is there a way to see the contents of the libraries included in the JAR? I would assume yes, but I don't really know how it's packaged.
